Framework/Geometry/test/BeamlineRayTracerTest.h

```diff
@@ -0,0 +1,774 @@
+    TS_ASSERT_DELTA(intersect.exitPoint.X(), 0.01, 1e-6);
+    TS_ASSERT_DELTA(intersect.exitPoint.Y(), 0.0, 1e-6);
+    TS_ASSERT_DELTA(intersect.exitPoint.Z(), 5.003464, 1e-6);
+
+    // Component that should have been intersected
+    TS_ASSERT_EQUALS(m_compInfo->indexOf(intersect.componentID),
+                     interceptedPixel);
+  }
+
+  /**
+   * Test ray tracing into a outlined detector.
+   *
+   * @param spec :: Parameters for the ray to test
+   */
+  void doTestOutlinedDetector(const RayTraceTestSpec &spec) {
+
+    // Force to be unit vector
+    auto testDir = spec.beamDirection;
+    testDir.normalize();
+
+    // Do a trace and store the results
+    auto results = RayTracer::traceFromSample(testDir, *m_compInfoOutlined);
+
+    // Expect no intersection
+    if (spec.pixelX == -1) {
+      TSM_ASSERT_LESS_THAN(spec.message, results.size(), 2);
+      return;
+    }
+
+    // Check size is correct, otherwise quit
+    TSM_ASSERT_EQUALS(spec.message, results.size(), 2);
+    if (results.size() < 2) {
+      TS_FAIL("Did not hit a detector when we should have.");
+      return;
+    }
+
+    // Get the first result
+    auto res = *results.begin();
+
+    // Check if what we found was a detector
+    const auto detIndex = m_compInfoOutlined->indexOf(res.componentID);
+    if (!m_compInfoOutlined->isDetector(detIndex)) {
+      TS_FAIL("Expected a detector but found none");
+      return;
+    }
+
+    const auto &det = m_detInfoOutlined->detector(detIndex);
+
+    // Find the xy index from the detector ID
+    TSM_ASSERT_EQUALS(spec.message, det.getPos().X(), spec.pixelX);
+    TSM_ASSERT_EQUALS(spec.message, det.getPos().Y(), spec.pixelY);
+  }
+
+  /**
+   * Test ray tracing into a rectangular detector.
+   *
+   * @param spec :: Parameters for the ray to test
+   */
+  void doTestRectangularDetector(const RayTraceTestSpec &params) {
+
+    // Force to be unit vector
+    auto testDir = params.beamDirection;
+    testDir.normalize();
+
+    // Do a trace and store the results
+    auto results = RayTracer::traceFromSample(testDir, *m_compInfoRectangular);
+
+    // Expect no intersection
+    if (params.pixelX == -1) {
+      TSM_ASSERT_LESS_THAN(params.message, results.size(), 2);
+      return;
+    }
+
+    // Check size is correct, otherwise quit
+    TSM_ASSERT_EQUALS(params.message, results.size(), 2);
+    if (results.size() < 2)
+      return;
+
+    // Get the first result
+    auto res = *results.begin();
+
+    // Get the detector
+    const auto detIndex = m_compInfoRectangular->indexOf(res.componentID);
+    if (!m_compInfoRectangular->isDetector(detIndex)) {
+      TS_FAIL("Expected a detector but found none");
+      return;
+    }
+
+    const auto &det = m_detInfoRectangular->detector(detIndex);
+    const auto columnIndex = m_compInfoRectangular->parent(detIndex);
+    const auto bankIndex = m_compInfoRectangular->parent(columnIndex);
+
+    const auto bank = dynamic_cast<const RectangularDetector *>(
```

Comment:
I think this could just become `componentInfo.type()` if you want a type check

instead of `bank->getXYForDetectorID(det.getID());` can just get the position directly and look at x,y. `componentInfo.position(detectorIndex)`.

Reply:
There's a difference here. `getXYForDetectorID` is check the row/column index of the rectangular detector not the position. Perhaps a better name is `getRowColForDetector`?

Reply:
Ok, I see. I would still rather this was done without `Instrument 1.0` I think we can establish column  and row index using the `children` and other related functions on `ComponentInfo`. This could go in to a helper function in the test, or better yet into https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/inc/MantidGeometry/Instrument/ComponentInfoBankHelpers.h

Reply:
👍 Agreed.